blockchain/message_pool/src/msgpool.rs

```diff
@@ -776,11 +877,303 @@ where
+            valid: true,
+            merged: false,
+        };
+        MsgChain {
+            index: 0,
+            chain: vec![node],
+        }
+    };
+
+    let mut chains = Vec::new();
+    let mut cur_chain = MsgChain::new();
+
+    for (i, m) in msgs.into_iter().enumerate() {
+        if i == 0 {
+            cur_chain = new_chain(m, i);
+            continue;
+        }
+        let gas_reward = cur_chain.curr().unwrap().gas_reward.clone() + &rewards[i];
+        let gas_limit = cur_chain.curr().unwrap().gas_limit + m.gas_limit();
+        let gas_perf = get_gas_perf(&gas_reward, gas_limit);
+
+        // try to add the message to the current chain -- if it decreases the gasPerf, then make a
+        // new chain
+        if gas_perf < cur_chain.curr().unwrap().gas_perf {
+            chains.push(cur_chain.clone());
+            cur_chain = new_chain(m, i);
+        } else {
+            let cur = cur_chain.curr_mut().unwrap();
+            cur.msgs.push(m);
+            cur.gas_reward = gas_reward;
+            cur.gas_limit = gas_limit;
+            cur.gas_perf = gas_perf;
+        }
+    }
+    chains.push(cur_chain);
+
+    // merge chains to maintain the invariant
+    loop {
+        let mut merged = 0;
+        for i in (1..chains.len()).rev() {
+            let (head, tail) = chains.split_at_mut(i);
+            if tail[0].curr().unwrap().gas_perf >= head.last().unwrap().curr().unwrap().gas_perf {
+                let mut chain_a_msgs = tail[0].curr().unwrap().msgs.clone();
+                head.last_mut()
+                    .unwrap()
+                    .curr_mut()
+                    .unwrap()
+                    .msgs
+                    .append(&mut chain_a_msgs);
+                head.last_mut().unwrap().curr_mut().unwrap().gas_reward +=
+                    &tail[0].curr().unwrap().gas_reward;
+                head.last_mut().unwrap().curr_mut().unwrap().gas_limit +=
+                    head.last().unwrap().curr().unwrap().gas_limit;
+                head.last_mut().unwrap().curr_mut().unwrap().gas_perf = get_gas_perf(
+                    &head.last().unwrap().curr().unwrap().gas_reward,
+                    head.last().unwrap().curr().unwrap().gas_limit,
+                );
+                tail[0].curr_mut().unwrap().valid = false;
+                merged += 1;
+            }
+        }
+        if merged == 0 {
+            break;
+        }
+        chains.retain(|c| c.curr().unwrap().valid);
+    }
+    // No need to link the chains because its linked for free
+    Ok(chains)
+}
+pub fn get_gas_reward(msg: &SignedMessage, base_fee: &BigInt) -> BigInt {
+    let mut max_prem = msg.gas_fee_cap() - base_fee;
+    if &max_prem < msg.gas_premium() {
+        max_prem = msg.gas_premium().clone();
+    }
+    max_prem * msg.gas_limit()
+}
+pub fn get_gas_perf(gas_reward: &BigInt, gas_limit: i64) -> f64 {
+    let a = gas_reward * types::BLOCK_GAS_LIMIT;
+    a.to_f64().unwrap() / gas_limit as f64
```

Comment:
this isn't safe? Why is it assumed that f64 can accommodate the value?

Reply:
https://doc.rust-lang.org/std/f64/constant.MAX.html
MAX is stupidly high

Reply:
OK, so i did some research. Im using the num-rational crate now here. It has the ability to convert a ratio of integers to a f64. I unwrap because it seems the only case where it returns `None` is when the denominator is 0. This should never be the case because the `gas_limit` is guaranteed to have a non-zero value.